packages/astro/src/vite-plugin-build-html/index.ts

```diff
@@ -253,17 +259,46 @@ export function rollupPluginAstroBuildHTML(options: PluginOptions): VitePlugin {
         assetIdMap.set(assetPath, referenceId);
       }
 
+      // Sort CSS in order of appearance in HTML
+      // This is needed because bundling is an async process where the chunks may
+      // be processed in any order. So we have to do a little bit of work to determine
+      // chunk order based off of HTML scanning in options() earlier
+      const sortedCSSChunks = [...pureCSSChunks];
+      sortedCSSChunks.sort((a, b) => {
+        let aIndex = Math.min(
+          ...Object.keys(a.modules).map((id) => {
+            const i = pageStyleImportOrder.findIndex((url) => id.endsWith(url));
+            return i >= 0 ? i : Infinity;
+          })
+        );
+        let bIndex = Math.min(
+          ...Object.keys(b.modules).map((id) => {
+            const i = pageStyleImportOrder.findIndex((url) => id.endsWith(url));
+            return i >= 0 ? i : Infinity;
+          })
+        );
+        return aIndex - bIndex;
+      });
+      const sortedChunkNames = sortedCSSChunks.map(({ fileName }) => fileName);
+
       // Create a mapping of chunks to dependent chunks, used to add the proper
       // link tags for CSS.
-      for (const chunk of pureCSSChunks) {
-        const chunkReferenceIds: string[] = [];
-        for (const [specifier, chunkRefID] of chunkToReferenceIdMap.entries()) {
-          if (chunk.imports.includes(specifier) || specifier === chunk.fileName) {
-            chunkReferenceIds.push(chunkRefID);
-          }
+      for (const chunk of sortedCSSChunks) {
+        const chunkModules = [chunk.fileName, ...chunk.imports];
+        chunkModules.sort((a, b) => {
```

Comment:
The next sort is within each chunk: we take each chunk’s dependencies, and sort them according to the “global ordering” above. This produces consistent ordering. It’s this final sorting that determines the build HTML ordering (which, again, will hopefully match the original ordering collected in the beginning of this file before bundling)